### DISCUSSION TITLE
How to avoid kernel crashes in Jupyter with reasonably large datasets?

### DISCUSSION THREAD

**Discussion by JosephKuchar** [moj-analytical-services/splink#2220] [Q&A]
So I'm Trying to use Splink to compare two datasets, DF1 and DF2. DF1 has ~100,000 records while DF2 has ~10,000 records. These are both pandas dataframes. Ultimately what I'm trying to do is not to get likelihood of a match (because the dataframes don't contain exactly the same information), but I am trying to extract the gamma values for my custom comparisons for the 'best' possible matches. 

My settings dictionary is the following:

```
settings = {
"link_type": "link_only",
'probability_two_random_records_match':0.01,
"blocking_rules_to_generate_predictions": [
    block_on("Country")
],
"comparisons": [
    comparison_name_1,
    comparison_name_2,
    comparison_name_3,
    comparison_date,
    comparison_address_1,
    comparison_address_2
],
"retain_intermediate_calculation_columns": True,
"retain_matching_columns":True,
}
```
It's possible the blocking variable I've chosen doesn't do enough to reduce pairs, as one country is far and away the most common. Unfortunately, the nature of the data makes it hard to block with much more precision than that.

When I try to generate predictions, I tend to get a kernel crash in the Jupyter notebook that I'm running it in. I managed in the end to generate what I need by iterating through chunks and restarting at checkpoints when the kernel crashed, but it's obviously not much of a solution:

```
list_df = np.array_split(DF1, 10)
i=0
for df in list_df[i:i+1]:
    
    linker = DuckDBLinker([df, DF2], settings)
    test=linker.predict().as_pandas_dataframe()
```
The machine itself has several hundred gigs of RAM, so I don't think kernel crashes are inevitable here. My first thought is that I'm feeding it dataframes - would it be better to import the data first into a duckdb database so that it's on disk rather than in memory? And if so, how do I do that? Another option is to run it from the terminal so that Jupyter's throttling doesn't impact it, but if there are other options I'd love to hear them!

Thanks,
Joseph

**Comment by RobinL** [comment]
How many records are there in the largest country (if you do a `select country, count(*) from df group by country` what's the biggest one?)

In terms of loading into duckdb, you would 
- create a duckdb connection to an on-disk db
- use a `create table mytable as (select * from read_parquet("your_data"))`
- then pass the string `"mytable"` into the DuckDB linker in place of `df`

You could also consider setting 
```
"retain_intermediate_calculation_columns": False,
"retain_matching_columns":False,
```

But given you've loads of memory i'm not sure how much that would help

If you really have to split up the job into multiple linkers one way to do it is by creating a random int column with values of say 1 to 10.

Then you can run 10 separate links with blocking rules like 
Job 1: l.country=r.country and l.salt = 1
Job 2: l.country=r.country and l.salt = 1

I'm not sure whether that would end up being much different to your current approach but maybe worth a try

Another thing you could try is passing a salting value to the blocking rule:

`block_on("Country", salting_partitions=4)`

Which will potentially/possibly make the job parallelise better

Finally - one blocking trick for things like addresses is to derive columns called things like 'unusual_token_1' 'unusual_token_2' and then use those for blocking purposes.  There's an example of this approaches using address data here:

https://github.com/RobinL/uk_address_matcher
In particular take a look at the format of 
`df_1_c = clean_data_using_precomputed_rel_tok_freq(df_1, con=con)`
![image](https://github.com/moj-analytical-services/splink/assets/2608005/d61c87ac-05e4-439a-95b7-05a3ce720580)

**Comment by JosephKuchar** [comment]
Thanks for the suggestions! I did try turning off the intermediate calculations columns, but ultimately what I need are the gamma columns, so they need to be retained. 

The country with the most records contains about 70K, so it's far and away the most records and really not ideal for blocking. The issue here is that specific addresses are hard to block on (my custom address comparisons are essentially combining multiple comparisons with OR statements, so "address_1_l matches address_1_r OR address_2_l matches address_1_r OR...) because I'm looking for the same entities to be present, but there are no hard rules about exactly which field they need to be present in.

Ultimately I've sorted it for now by decreasing the chunk size by another factor of 10 - the in-use memory dropped from over 100G to about 10G, which is much more reasonable. But I think I will also try the approach of loading the data into a local duckdb database and let you know how that goes!
